src/lib.rs

```diff
@@ -258,6 +258,17 @@ fn parse_braced_variable(source: &[u8], finger: usize) -> Result<Variable, Error
 	})
 }
 
+fn get_maybe_char_at(data: &[u8], index: usize) -> error::CharOrByte {
+	if let Ok(string) = std::str::from_utf8(data) {
+		if string.is_char_boundary(index) {
+			if let Some(c) = string[index..].chars().next() {
+				return error::CharOrByte::Char(c);
+			}
+		}
+	}
+	error::CharOrByte::Byte(data[index])
+}
```

Comment:
The current implementation can't seem to handle the case where the source aren't all valid UTF-8 but the bytes at `index` are.

```rust
fn get_maybe_char_at(data: &[u8], index: usize) -> error::CharOrByte {
	let data = &data[index..];
	let head = &data[..data.len().min(4)]; // an UTF-8 encoded char is 4 bytes max
	assert!(!head.is_empty());

	match std::str::from_utf8(head) {
		Ok(s) => {
			// All four bytes are valid char(s), return the first one
			error::CharOrByte::Char(s.chars().next().unwrap())
		},
		Err(err) => {
			if err.valid_up_to() > 0 {
				// At least one valid char
				let valid_char_bytes = &head[..err.valid_up_to()];
				error::CharOrByte::Char(std::str::from_utf8(valid_char_bytes).unwrap().chars().next().unwrap())
			} else {
				// First byte isn't valid char
				error::CharOrByte::Byte(head[0])
			}
		}
	}
}
```

Reply:
Ah, good catch. Shame there isn't an `std` function that directly returns the valid portion as `str`.

I've adopted your fix, although slightly tweaked.